src/transformStateWithClones.js

```diff
@@ -7,7 +7,43 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const actionsArray = [];
+  const resultObject = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(resultObject, action.extraData, actionsArray);
+        break;
+      case 'removeProperties':
+        removeProperties(resultObject, action.keysToRemove, actionsArray);
+        break;
+      case 'clear':
+        clear(actionsArray, resultObject);
```

Comment:
You're directly manipulating the `resultObject` inside the switch cases, which goes against the checklist point about not mutating objects. Instead, you should create a new object copy within each case.